library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
library(clusterProfiler)
library(ReactomePA)
#setwd('/WORK/lbyybl/WH/rvb/chip-seq/PolII/unique/diff_peak/class')
setwd('/WORK/lbyybl/WH/rvb/cor_uniq/sample_Ruvbl/Pol2/diff_peak/overlap')
annopeak_fuc <- function(peak_fle,name){
  peak <- readPeakFile(peak_fle)
  # pdf(paste0(name,'peak_chr.pdf'),width = 6,height = 5)
  # covplot(peak_fle,chrs = c(paste0('chr',c(1:19,'X','Y'))))
  # dev.off()
  # 
  # pdf(paste0(name,'promoter_hetmap.pdf'),width = 3,height = 7)
  # peakHeatmap(peak_fle, TxDb=txdb, upstream=3000, downstream=3000, color="red")
  # dev.off()
  # 
  # pdf(paste0(name,'promoter_profile.pdf'),width=5,height = 5)
  # plotAvgProf2(peak_fle, TxDb=txdb, upstream=3000, downstream=3000,
  #              xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
  # dev.off()
  
  peakAnno <- annotatePeak(peak_fle, tssRegion=c(-2000, 2000),
                           TxDb=txdb, annoDb="org.Mm.eg.db")
  # colors   <- rainbow(10)
  # pdf(paste0(name,'pieplot_anno.pdf'),width = 6,height = 4)
  # plotAnnoPie(peakAnno)
  # dev.off()
  # pdf(paste0(name,'bar_anno.pdf'),width = 6,height = 3)
  # plotAnnoBar(peakAnno)
  # dev.off()
  # pdf(paste0(name,'venn_anno.pdf'),width = 5,height = 5)
  # vennpie(peakAnno)
  # dev.off()
  # pdf(paste0(name,'upset_anno.pdf'),width = 1,height = 1)
  # upsetplot(peakAnno)
  # dev.off()
  # pdf(paste0(name,'unset_venn_anno.pdf'),width = 12,height = 12)
  # upsetplot(peakAnno, vennpie=TRUE)
  # dev.off()
  # pdf(paste0(name,'dis2tss_anno.pdf'),width = 6,height = 3)
  # plotDistToTSS(peakAnno,
  #               title="Distribution of transcription factor-binding loci\nrelative to TSS")
  # dev.off()
  #library(ReactomePA)
  
  pathway1 <- enrichPathway(as.data.frame(peakAnno)$geneId,organism = 'mouse')
  head(pathway1, 2)
  gene <- seq2gene(peak, tssRegion = c(-1000, 1000), flankDistance = 3000, TxDb=txdb)
  enrich_gene <- data.frame('name'=gene)
  fwrite(enrich_gene,paste0(name,'enrich_gene.txt'),col.names = F)
  pathway2 <- enrichPathway(gene,organism = 'mouse')
  head(pathway2, 2)
  pdf(paste0(name,'seq2gene_enrich.pdf'),width = 30,height = 5)
  print(dotplot(pathway2))
  #print(plot)
  dev.off()
  pdf(paste0(name,'enrich.pdf'),width=30,height = 5)
  print(dotplot(pathway1))
  #print(plot1)
  dev.off()
  
}
for (file in system('ls *.bed',intern = T)){
  #print(file)
  annopeak_fuc(file,strsplit(file,"\\.")[[1]][1])
}

# peak_fle <- '../unchange_deseq2.txt'
# name <- 'unchange'
# annopeak_fuc('../r1_high_deseq2.txt','Ruvbl1')
# annopeak_fuc('../r2_high_deseq2.txt','Ruvbl2')
# annopeak_fuc('../unchange_deseq2.txt','unchange')
# peak <- readPeakFile('r1_high_deseq2.txt')
# pdf('peak_chr.pdf',width = 6,height = 5)
# covplot(peak,chrs = c(paste0('chr',c(1:19,'X','Y'))))
# dev.off()
# 
# pdf('promoter_hetmap.pdf',width = 3,height = 7)
# peakHeatmap('r1_high_deseq2.txt', TxDb=txdb, upstream=3000, downstream=3000, color="red")
# dev.off()
# 
# pdf('promoter_profile.pdf',width=5,height = 5)
# plotAvgProf2('r1_high_deseq2.txt', TxDb=txdb, upstream=3000, downstream=3000,
#              xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
# dev.off()
# 
# peakAnno <- annotatePeak('r1_high_deseq2.txt', tssRegion=c(-2000, 2000),
#                          TxDb=txdb, annoDb="org.Mm.eg.db")
# colors   <- rainbow(10)
# pdf('pieplot_anno.pdf',width = 6,height = 4)
# plotAnnoPie(peakAnno)
# dev.off()
# pdf('bar_anno.pdf',width = 6,height = 3)
# plotAnnoBar(peakAnno)
# dev.off()
# pdf('venn_anno.pdf',width = 5,height = 5)
# vennpie(peakAnno)
# dev.off()
# pdf('upset_anno.pdf',width = 1,height = 1)
# upsetplot(peakAnno)
# dev.off()
# pdf('unset_venn_anno.pdf',width = 12,height = 12)
# upsetplot(peakAnno, vennpie=TRUE)
# dev.off()
# pdf('dis2tss_anno.pdf',width = 6,height = 3)
# plotDistToTSS(peakAnno,
#               title="Distribution of transcription factor-binding loci\nrelative to TSS")
# dev.off()
# #library(ReactomePA)
# 
# pathway1 <- enrichPathway(as.data.frame(peakAnno)$geneId,organism = 'mouse')
# head(pathway1, 2)
# gene <- seq2gene(peak, tssRegion = c(-1000, 1000), flankDistance = 3000, TxDb=txdb)
# enrich_gene <- data.frame('name'=gene)
# fwrite(enrich_gene,'enrich_gene.txt',col.names = F)
# pathway2 <- enrichPathway(gene,organism = 'mouse')
# head(pathway2, 2)
# pdf('seq2gene_enrich.pdf',width = 8,height = 5)
# dotplot(pathway2)
# dev.off()
# pdf('enrich.pdf',width=7,height = 5)
# dotplot(pathway1)
# dev.off()
